### ISSUE TITLE
Consider reformatting help example code

### ISSUE THREAD

**Issue by prowley** [premake/premake-core#1548]
**What problem will this solve?**
Confusion.

Throughout the example code indents are used as a stylized format for the example code snippets. As someone new to remake this was highly misleading because indent in a language with no other scoping indicators generally means scope (and even where there are indicators, any whitespace is usually inserted to reinforce some existing scope to the reader). I appreciate that there are places where scope is explained, however even then, it is 3 sections in, and having seen previous examples there is confusion around what exactly is meant by scope in the explanation. Ultimately, in premake5 scope seems to mean "anything after this" unless another thing like this one is declared. Nothing wrong with that, but the indents in the examples do not help to communicate that. I didn't really understand the explanation of scope until I spent 4 hours messing around with a test project that had, at the time, inexplicable outcomes. And the confusion I had can be directly traced back to those indents.

**What might be a solution?**
Forgoing indents in the examples entirely would be best, since then "cherry picking" the docs would make clear there are no indented scopes, and if that were surprising, might prompt further doc investigation.

**What other alternatives have you already considered?**
Something which explains the basic scoping in the first example, at that point in the doc. If it said right away "all these ident are meaningless, here's how the scope works here" I'd be 4 hours younger :) This wouldn't catch hardcore cherry pickers, but most will at least skim the first example and text around it.

**Anything else we should know?**
It actually works nicely, and quite intuitively, once the actual scoping mechanism is correctly understood, it's just getting to the correctly part .

**Comment by nickclark2016** [comment]
It's probably just because I'm very used to working with Premake and Lua, but could you give a snippet for how you would recommend formatting the example?

**Comment by samsinsane** [comment]
We're not changing the formatting of the snippets. If you believe that concepts in [Your First Script](https://github.com/premake/premake-core/wiki/Your-First-Script) should be introduced earlier, I'm open to hearing about that, but to suggest that we write our docs for people who don't read them is ridiculous.

**Comment by prowley** [comment]
Let me address this up front: "but to suggest that we write our docs for people who don't read them is ridiculous."

To counter I would suggest that thinking everybody who uses premake should read the docs from start to finish, adjusting their understanding of things as they read each section, is something which does not match the reality how people use tech docs.

People read tech docs in different ways according to experience (emphasis on read, not don't read, if I hadn't read your docs, I wouldn't have bothered with this issue either - I'm kinda busy.) Common among those ways is to look at a few examples and maybe some surrounding text, decide you understand the gist of it, then refer to api references, often referring back to prose sections to get better insight on some topic. If when I first read K&R, it had formatted every example in some arbitrary manner that didn't actually match the code structure, I'm sure it would have been a lot harder to get through and out the other side knowing C.

Perhaps I missed it, but Your First Script actually has all the ingredients I mentioned in the issue for creating misunderstandings about scope, but none of the explanation about scope. 

I must admit when I first read it I started to tune out and skip at What's going On Here, and tune back in to the things not telling me about lua and how it works. I'm not sure the lua explanation is even necessary at that point, but certainly should be somewhere in there. Anyway, a bullet point in that section is:

The indentation whitespace is arbitrary; this is the way I happen to like it.

followed by two more bullets telling me how Lua works. Honestly, if that bullet list were the only thing in that subsection it would be better. FYI I don't read chapters or sections in books about a language when I already know it either, unless that book happens to be about the language itself. This section kinda reads like that sort of preamble, but buries the message. The problem is that message is more important than pretty much anything around it for understanding scope in the example, and even then it doesn't actually explain much at all about scope.

Now look at the script, with an eye to indentation as a communication device, not just something to make the source code look pretty.

```
-- premake5.lua
workspace "HelloWorld"
   configurations { "Debug", "Release" }

project "HelloWorld"
   kind "ConsoleApp"
   language "C"
   targetdir "bin/%{cfg.buildcfg}"

   files { "**.h", "**.c" }

   filter "configurations:Debug"
      defines { "DEBUG" }
      symbols "On"

   filter "configurations:Release"
      defines { "NDEBUG" }
      optimize "On"
```

The minimal change here to make the indentation make sense would be to indent project too, since it is within the scope of workspace (if the project was in another file, you would indent the include, right?). Understanding scope for this example is not important, but _misunderstanding_ the scope from this example _is_ important. So since it is introduced by side-channel, it really ought to be explained.

Here are the key scoping concepts I think should probably be up front if all the examples are to be formatted in this way (and let's be honest I'm a day or two in to premake, and I've being trying to build something during that time while also figuring out this stuff, so it's quite possible I have things wrong here. But then that would merely reinforce my point, this isn't my first rodeo by a long shot):

1. Anything in a workspace is globally inherited
2. Projects only inherit from the workspace, never other projects, specifically _linked_ sub-projects inherit nothing from the project linking
3. Filters have scope until the next filter declaration (this is mentioned elsewhere, but again, the indentation, which suggests an if..then structure, and therefore an end of scope too) The takeaway here is you will eventually get side-effects of this rule if you don't declare all the non filter stuff before the first filter (adding filter {} at the end is a bit of a hack imho, and will lead to issues when (easily) forgotten - if there were a "filterend" I'd feel better about it, since that actually mimics the suggested if..then..end, and a lint could actually detect possible issues) The API reference for filter does a great job of explaining this, but we're in example/explain territory here.

At least, that's my current understanding.

**Comment by prowley** [comment]
@nickclark2016 Re:how I would reformat. I'm not sure there is an ideal way tbh. Although everything is Lua under the covers, and you can write lua in the config, one must be mindful that the config file is really a DSL that happens to be able to include Lua-like syntax, so language design principals apply. I'm not even sure how scope and raw lua interacts with scoping, I haven't done anything like that yet, but I suspect it would look like the language scoping is obeyed, but there are gotchas hidden within because e.g. lua within filters would execute regardless of whether the filter applies, so side-effects would matter.

If I were to reformat the example to reflect how the language works rather than attempting to describe the scope it would be very much flatter, Makefile-like even. That would have the advantage, from a communication point of view, of showing that everything is executed consecutively, and some things just happen to change the behavior of the following things - like a Makefile. For me, that would prompt the question, "wait, filter? what's it filtering?", and after research the answer would be "oh, everything until the next filter."

I think to format in a way that suggests scope with indents, it would ideally have to have support for scoping indicators in the DSL, endfilter ,endproject etc., possibly, and preferably, with checks that the previous thing was ended before a new one starts - warnings would be sufficient.
Then for example, this would be both intuitive and not error prone:

```
filter "configurations:Debug"
  --do some stuff here that is pertinent to surrounding code
endfilter

-- do some stuff for everything

filter "configurations:Debug"
  --do some other stuff here that is pertinent to this section of code
endfilter
```

This would actually be my preference over a flat representation - I like seeing scope in indents, it helps me understand the structure. Probably most importantly, a linter could tell me the code might not be doing what I think it is if I leave off the endfilter.

**Comment by nickclark2016** [comment]
`endfilter` is not in Premake.  If you want to end a filter, you start a new filter.  If you want to disable the active filter rather than starting a new one, you use `filter {}`.  To make it more lua-like, it could be useful to add an `endfilter` piece to the API, which just aliases to `filter {}`, but that is out of the scope of the pull request as initially defined.  @samsinsane Would you consider adding the `filter {}` statements to the end of the filtered blocks to be an appropriate response to the PR, as it does raise a fair point from the perspective of a beginner to Premake.  I think that an `endfilter` piece deserves further discussion.

**Comment by starkos** [comment]
We've been using this format since Premake4 was released, which is long enough ago now that I have no idea how long ago it was. Seven years? Ten years? A long time.

FYI, I've already [implemented and proposed a new syntax to make scoping explicit for a future version](https://opencollective.com/premake/updates/community-update-6). It may be possible to backport to v5.

```
workspace('MyWorkspace', function()
   configurations { 'Debug', 'Release' }

   project('MyProject', function()
      kind 'ConsoleApp'
      files { '**.h', '**.c' }
   end)
end)
```

@nickclark2016 Okay if I let you run with this one?

**Comment by nickclark2016** [comment]
Running with a backport of scoping or with updating the example?

**Comment by starkos** [comment]
Your call. I think the options are adding an `endfilter()` or back porting. The former would certainly be easier, IMHO.

**Comment by nickclark2016** [comment]
I'd argue let's just add an endfilter(), then worry about backporting later if needed.  While I absolutely love the updated syntax, I'd rather see that in a premake6 than try to retrofit it back onto premake5.

**Comment by prowley** [comment]
FWIW I really like @starkos solution, it's not just Lua like, it's lua, and solves scoping completely i.e. it's true scoping that will hold up to custom lua code.

**Comment by nickclark2016** [comment]
@prowley Very much agree, but I don't want to introduce bifurcation of the filtering mechanism at this point in time for Premake 5.

**Comment by nickclark2016** [comment]
I'm finishing up implementing `endfilter()` now, which delegates to `filter {}`

**Comment by starkos** [comment]
@samsinsane - if you're around, I wouldn't mind getting your input on this. I'm a little uneasy about introducing a new convention into the new user documentation when we already know we're going to change it. Better to just bite the bullet and backport the function approach? Or leave as-is?

**Comment by samsinsane** [comment]
In the six or so years that I've used Premake, I haven't seen anyone suggest that something like `endfilter()` needs to exist. So, I'm not really sure what problem is trying to be solved here? To me, `endfilter()` is inconsistent, and will create confusion. Inconsistent because it's the only API to my knowledge that requires parenthesis, and confusing because you don't end a project or workspace, only a filter.

On top of this, there are questions that don't seem to be answered for me:
- Will `endfilter()` become mandatory?
- Does `endfilter()` replace `filter {}`? Will using `filter {}` emit a deprecation warning to use `endfilter()` instead?
- What happens if a user overrides `endfilter` or `filter`? What if they want to override both, or just one?
- Will we need to create `endproject()` and `endworkspace()` for consistency? Same questions above apply to these too.

Overall, I'm against `endfilter()` if we're going to end up with a different convention anyway. I'd rather we focused on fixing the issues with the new, and clearly superior, convention than trying to improve what we have. It simplifies the transition as well by reducing the amount of documentation required to guide people through upgrading their scripts.

---

Regarding the question of backporting. Premake5 is in alpha, so as long as we continue to support the existing system and emit a deprecation warning, then I don't see any issues with backporting? Is it possible to optionally use the new system? For example, if a user had to call `enableExperimentalFormat()` at the top of the `premake5.lua`, would it be feasible to replace the current code path with the new code path, while also still allowing the user to update scripts slowly? Something like this:
```lua
enableExperimentalFormat()

workspace 'MyWorkspace'
  configurations { 'Debug', 'Release' }

  project 'MyLib'
    kind 'StaticLib'
    files { '**.h', '**.c' }

  project('MyProject', function()
    kind 'ConsoleApp'
    links { 'MyLib' }
    files { '**.h', '**.c' }
  end)
```

I haven't really thought through this that much, it was just an idea I had as I was replying. I imagine there's a lot of edge cases to be considered and perhaps a clean break is required.

**Comment by redorav** [comment]
Without knowing too much the internals of premake , would it be possible to have a simpler scope using {}? Something like:

project('MyProject',
{
    kind 'ConsoleApp'
    links { 'MyLib' }
    files { '**.h', '**.c' }
 })

I find the function() and end part a bit verbose, just distracts from what the code is really trying to say.

**Comment by nickclark2016** [comment]
> 
> 
> In the six or so years that I've used Premake, I haven't seen anyone suggest that something like `endfilter()` needs to exist. So, I'm not really sure what problem is trying to be solved here? To me, `endfilter()` is inconsistent, and will create confusion. Inconsistent because it's the only API to my knowledge that requires parenthesis, and confusing because you don't end a project or workspace, only a filter.
> 
> On top of this, there are questions that don't seem to be answered for me:
> 
>     * Will `endfilter()` become mandatory?
> 
>     * Does `endfilter()` replace `filter {}`? Will using `filter {}` emit a deprecation warning to use `endfilter()` instead?
> 
>     * What happens if a user overrides `endfilter` or `filter`? What if they want to override both, or just one?
> 
>     * Will we need to create `endproject()` and `endworkspace()` for consistency? Same questions above apply to these too.
> 
> 
> Overall, I'm against `endfilter()` if we're going to end up with a different convention anyway. I'd rather we focused on fixing the issues with the new, and clearly superior, convention than trying to improve what we have. It simplifies the transition as well by reducing the amount of documentation required to guide people through upgrading their scripts.
> 
> Regarding the question of backporting. Premake5 is in alpha, so as long as we continue to support the existing system and emit a deprecation warning, then I don't see any issues with backporting? Is it possible to optionally use the new system? For example, if a user had to call `enableExperimentalFormat()` at the top of the `premake5.lua`, would it be feasible to replace the current code path with the new code path, while also still allowing the user to update scripts slowly? Something like this:
> 
> ```lua
> enableExperimentalFormat()
> 
> workspace 'MyWorkspace'
>   configurations { 'Debug', 'Release' }
> 
>   project 'MyLib'
>     kind 'StaticLib'
>     files { '**.h', '**.c' }
> 
>   project('MyProject', function()
>     kind 'ConsoleApp'
>     links { 'MyLib' }
>     files { '**.h', '**.c' }
>   end)
> ```
> 
> I haven't really thought through this that much, it was just an idea I had as I was replying. I imagine there's a lot of edge cases to be considered and perhaps a clean break is required.

I'm there with you. I've never heard of a need for `endfilter`.  You raise a lot of good points regarding how we'd handle overriding the functions.  For now, I'll hold off on any work for this until we have a path forward.

**Comment by starkos** [comment]
> Is it possible to optionally use the new system?

Yes. It could be done something like (pseudocode obviously)…

```lua
function when(terms, fn)
   api.pushState()  -- need to figure this out
   filter(terms)
   fn()
   api.popState()  -- need to figure this out
end
```

This is compatible with the existing approach, so you could drop it in anywhere, no need to choose between the approaches. Can migrate incrementally.

If we go this route, I think it makes sense to also switch over the wiki documentation so new users ramp up on new syntax instead of the old. 

Probably not hard to do. I'll try to get a look at it.

FYI, for v6 (or whatever; really need to get a decision on that) I'd prefer to make the function scope the One True Way. I don't like having to juggle multiple ways of doing the same thing.

**Comment by samsinsane** [comment]
@redorav The issue with what you've suggested is that those functions execute before they're put in the array. Using your example of:
```lua
project('MyProject',
{
    kind 'ConsoleApp'
    links { 'MyLib' }
    files { '**.h', '**.c' }
})
```
The execution order will be `kind` -> `links` -> `files` -> `project`. You can see that if you run this:
```lua
function TestFunc1(prjname, array)
  print("TestFunc1:", prjname, #array, array)
end

function TestFunc2(val)
  print("TestFunc2:", val)
end

TestFunc1("MyProject",
{
  TestFunc2("test1");
  TestFunc2("test2");
})
```

You'll get something like this:
```
TestFunc2:	test1
TestFunc2:	test2
TestFunc1:	MyProject	0	table: 0x190feb0
```

This means that you're applying `kind`, `links` and `files` to your `workspace` or the previous `project`/`filter` instead of this new `project`.

---

> This is compatible with the existing approach, so you could drop it in anywhere, no need to choose between the approaches. Can migrate incrementally.

I guess the only issue is that this isn't how it would work in the new system right? In the new system the functions are only executed if necessary right? So for the above, if `terms` is not met, then `fn` is never executed? I feel like this difference could create problems, if you depend on a function to be executed even if the condition is false for example, but in v6 it won't be executed. It's not the end of the world, but I think changing this later might be confusing for some people and maybe backporting isn't a good idea because of this? I'm not sure.

> FYI, for v6 (or whatever; really need to get a decision on that) I'd prefer to make the function scope the One True Way. I don't like having to juggle multiple ways of doing the same thing.

I completely agree.

**Comment by redorav** [comment]
@samsinsane You're absolutely right, my lua-fu isn't the best, however I've been experimenting and I came up with two other alternatives, see what you think of this.

**(A)**

```lua
function a(thestring)
	print(thestring);
end

function b(thestring)
	print(thestring);
end

local c = 
{
	{ a, "hello" },
	{ b, "world" }
}

for _,f in ipairs(c) do
	f[1](f[2])
end
```

So the final syntax would become

```lua
project('MyProject',
{
    { kind, 'ConsoleApp' };
    { links, { 'MyLib' } };
    { files, { '**.h', '**.c' } };
})
```

It is noisier than the one I was after so that's a bit of a shame. 

Another alternative is to make kind, links, files, etc return a pair themselves, like

**(B)**

```
function kind(params)
    return { fn_kind, params }
end
```

so we could now write the original syntax as proposed, here for reference again:

project('MyProject',
{
    kind 'ConsoleApp';
    links { 'MyLib' };
    files { '**.h', '**.c' };
})

with the difference that now we'd need commas. Anyway, just my two cents there. The function approach is a lot more flexible in that you can run any code you want for the matching filter, it would be nice to just be able to write it inline somehow.

**Comment by starkos** [comment]
I think this issue should be closed: we're not going to un-indent the documentation. @samsinsane if you agree let's go ahead and shut this one down.

Another issue/RFC should be opened to discuss making scoping explicit. I don't have time to write that up this morning but I'll put it near the top of my list. Both approaches are doable. The table syntax is potentially cleaner, we might even be able to get rid of the trailing separator, but it triggers my [spidey senses](https://en.wiktionary.org/wiki/Spidey-sense). And the table syntax can't be easily backported to v5, while the function syntax can without excessive effort. ATM I prefer the function approach. 

@samsinsane - the function gets evaluated immediately; [a little more info here](https://opencollective.com/premake/updates/community-update-6).

**Comment by samsinsane** [comment]
@starkos I agree to closing this down and maybe opening another issue for discussing making the scoping explicit. Sorry for the confusion around the function execution, given that it's executed immediately, I don't think I have any concerns.

@redorav I'm not a fan of the array approach at all. It just feels super restrictive compared to what we have right now, and what @starkos has been working on.

**Comment by starkos** [comment]
@redorav I played around with your syntax more in preparation for writing up an RFC, but I couldn't come up with a great way to handle this:

```lua
when('configurations:Debug', {
    defines { 'DEBUG' };
    when('platforms:macOS', {
      defines { 'MACOS' }
   };
};
```

The inner `when()` could return a state table, which would then get processed as part of the outer `when()`. But the outer `when()` can't just return a state table, because there is no one to return it to. You would have to put  _everything_ in some top-level table, which would mean no longer being able to put arbitrary code or functions in your project scripts (since they can't go inside a table).

It was an idea worth considering but I think, based on the restrictions I'm seeing, I'm not going to bother opening an RFC. I think the function approach is the most obvious and understandable, and it isn't _that_ much noisier than the table when you take all the semicolons into account.

```lua
when('configurations:Debug', function()
   defines { 'DEBUG' }
   when('platforms:macOS', function()
      defines { 'MACOS' }
   end)
end)
```

Also: the function version is already done, _and_ it can be backported to v5 without great effort. I'm closing this issue as discussed, and adding "Backport explicit blocks to v5" to my to-do list. If someone can/wants to take it on sooner see [my comment above](https://github.com/premake/premake-core/issues/1548#issuecomment-724031401) and feel free to send questions my way.

Down the road, I'm not opposed to adding an alias for `function` to Premake's embedded Lua runtime (maybe `$()`?, or otherwise customizing the syntax

**Comment by redorav** [comment]
@starkos @samsinsane I know you've closed it but I'll leave this here just as a last reply, I think the function approach has a lot going for it and was just hypothesizing on possible equivalent solutions with different syntax. Clearly the function approach is more expressive and flexible, probably even performant, than the approaches I outlined so far, and I'm sure you've given it a lot more thought than I have, so it's fair enough and I totally understand and support the new model. If, on top of that, we can customize the syntax to create functions via a shorthand like you propose, I'd personally be extra happy for it.

Just as a FYI I found this while looking for shorthand Lua functions http://lua-users.org/wiki/ShortAnonymousFunctions, at the end they provide custom syntax that other interpreters have added, might be interesting to look at although $() would probably work great.
